src/chainlit/session.py

```diff
@@ -1,29 +1,70 @@
-from typing import Dict, TypedDict, Optional, Callable, Any, Union
+from secrets import token_urlsafe
+from typing import Dict, Optional, Callable, Any, Union
 from chainlit.client.base import BaseAuthClient, BaseDBClient
 from chainlit.types import AskResponse
 
 
-class Session(TypedDict):
+class Session:
     """Internal session object."""
 
-    # The ID of the session
-    id: str
-    # Function to ask the user a question
-    ask_user: Callable[[Any, Optional[int]], Union[AskResponse, None]]
-    # Function to emit a message to the user
-    emit: Callable[[str, Any], None]
-    # User specific environment variables. Empty if no user environment variables are required.
-    user_env: Dict[str, str]
-    # Optional langchain agent
-    agent: Any
-    # Optional llama instance
-    llama_instance: Any
-    # Whether the current task should be stopped
-    should_stop: bool
-    # Optional client to authenticate users
-    auth_client: Optional[BaseAuthClient]
-    # Optional client to persist messages and files
-    db_client: Optional[BaseDBClient]
-
-
-sessions: Dict[str, Session] = {}
+    sessions_id: Dict[str, str] = {}
+    sessions: Dict[str, "Session"] = {}
+
+    def __init__(
+        self,
+        # Session key
+        key: str,
+        # Function to emit a message to the user
+        emit: Callable[[str, Any], None],
+        # Function to ask the user a question
+        ask_user: Callable[[Any, Optional[int]], Union[AskResponse, None]],
+        # Optional client to authenticate users
+        auth_client: Optional[BaseAuthClient],
+        # Optional client to persist messages and files
+        db_client: Optional[BaseDBClient],
+        # User specific environment variables. Empty if no user environment variables are required.
+        user_env: Dict[str, str],
+        # Optional langchain agent
+        agent: Any = None,
+        # Optional llama instance
+        llama_instance: Any = None,
+    ):
+        self.key = key
+        self.ask_user = ask_user
+        self.emit = emit
+        self.user_env = user_env
+        self.agent = agent
+        self.llama_instance = llama_instance
+        self.auth_client = auth_client
+        self.db_client = db_client
+        self.should_stop = False
+        self.id = token_urlsafe()
+
+        Session.sessions[self.id] = self
+        Session.sessions_id[key] = self.id
+
+    def restore(self, new_key: str):
+        if self.key in Session.sessions_id:
+            Session.sessions_id.pop(self.key)
+        Session.sessions_id[new_key] = self.id
+
+    def delete(self):
+        if self.key in Session.sessions_id:
+            Session.sessions_id.pop(self.key)
+        Session.sessions.pop(self.id)
+
+    @classmethod
+    def get(self, key: str):
+        session_id = Session.sessions_id.get(key)
+        return Session.get_by_id(session_id)
```

Comment:
Wouldn't it make more sense to do the opposite, i.e. storing by key rather than id in the main store? We are going to look by key rather than id all the time except when we try to restore a session, aren't we?

Reply:
Also, why not use two session stores, one that's indexed by id and the other one by key, that both hold the references to the same session objects? That would avoid the intermediary step.